### PR TITLE
Use repository for missing service in 'bearer' auth challenge

### LINKED ISSUE #576: Authentication to Red Hat Container Registry (Openshift) doesn't work
**Description of the issue**:
Using token obtained from Red Hat Container Registry (RHCR) in settings.xml doesn't authenticate user correctly to registry.

**Expected behavior**:
Registry is authenticating user correctly.

**Steps to reproduce**: 
mvn compile jib:build with configuration set to point to RHCR and settings.xml containing valid credentials to that. According RHCR username is "unused", amd password is token provided by UI.

**Environment**: 
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T22:39:06+03:00)
Java version: 1.8.0_131, vendor: Oracle Corporation
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

jib-maven-plugin version 0.9.6

**`jib-maven-plugin` Configuration**: 
```xml
<configuration>
	<from>
		<image>openshift-redhat-registry/testProject/test-container</image>
	</from>
	<to>
		<image>openshift-redhat-registry/testProject/${project.artifactId}:${project.version}</image>
	</to>
	<container>
		<jvmFlags>
			<jvmFlag>-Xms512m</jvmFlag>
		</jvmFlags>
	</container>
</configuration>
```

**Log output**: 
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:build (default-cli) on project jib-test-project: Build image failed: Tried to retrieve authentication method for openshift-redhat-registry but failed because: Failed get authentication method from 'WWW-Authenticate' header | If this is a bug, please file an issue at https://github.com/GoogleContainerTools/jib/issues/new: Failed to authenticate with the registry because: 'service' was not found in the 'WWW-Authenticate' header, tried to parse: Bearer realm="https://openshift-redhat-registry/openshift/token" -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:build (default-cli) on project jib-test-project: Build image failed
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Build image failed
	at com.google.cloud.tools.jib.maven.BuildImageMojo.execute(BuildImageMojo.java:150)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: com.google.cloud.tools.jib.registry.RegistryErrorException: Tried to retrieve authentication method for openshift-redhat-registry but failed because: Failed get authentication method from 'WWW-Authenticate' header | If this is a bug, please file an issue at https://github.com/GoogleContainerTools/jib/issues/new
	at com.google.cloud.tools.jib.registry.RegistryErrorExceptionBuilder.build(RegistryErrorExceptionBuilder.java:108)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:104)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:31)
	at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:176)
	at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:139)
	at com.google.cloud.tools.jib.registry.RegistryClient.callRegistryEndpoint(RegistryClient.java:338)
	at com.google.cloud.tools.jib.registry.RegistryClient.getRegistryAuthenticator(RegistryClient.java:194)
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator$Initializer.initialize(RegistryAuthenticator.java:83)
	at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call(AuthenticatePushStep.java:85)
	at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call(AuthenticatePushStep.java:42)
	at com.google.common.util.concurrent.CombinedFuture$CallableInterruptibleTask.runInterruptibly(CombinedFuture.java:181)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.cloud.tools.jib.registry.RegistryAuthenticationFailedException: Failed to authenticate with the registry because: 'service' was not found in the 'WWW-Authenticate' header, tried to parse: Bearer realm="https://openshift-redhat-registry/openshift/token"
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.newRegistryAuthenticationFailedException(RegistryAuthenticator.java:150)
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.fromAuthenticationMethod(RegistryAuthenticator.java:141)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:98)
	... 13 more
```

**Additional Information**: 
Failure also occurs when trying:
1. Login with oc CLI tool
2. Using real username and token from `oc whoami -t` as password.

With direct call with Postman to https://openshift-redhat-registry/openshift/token authenticating with user _unused_ and given token, access_token is displayed correctly in response

With real username + that `oc whoami` -t -token that /token -endpoint returns 401.

### PR DESCRIPTION
**Pull request by briandealwis**
Fixes #576

### CODE REVIEW

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStepTest.java [hunk 1]
```diff
@@ -59,7 +59,7 @@
   private Path temporaryCacheDirectory;
 
   @Before
-  public void setUp() throws IOException, URISyntaxException {
```
Comment (briandealwis) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStepTest.java]: Sorry this is extraneous to this PR: this exception isn't actually thrown and flags a warning.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java [hunk 2]
```diff
@@ -137,10 +137,11 @@ static RegistryAuthenticator fromAuthenticationMethod(
 
     Pattern servicePattern = Pattern.compile("service=\"(.*?)\"");
     Matcher serviceMatcher = servicePattern.matcher(authenticationMethod);
-    if (!serviceMatcher.find()) {
-      throw newRegistryAuthenticationFailedException(authenticationMethod, "service");
+    // use the provided registry as the service (e.g., for OpenShift)
+    String service = repository;
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java]: I believe the service should actually be the `serverUrl` (https://docs.docker.com/registry/spec/auth/token/#how-to-authenticate)

Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java]: Example: 

`curl -v https://gcr.io/v2/ 2>&1 >/dev/null | grep "www-authenticate"`

Comment (briandealwis) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java]: The service will be replaced in the following two lines.  But let me rewrite it to be clearer.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java [hunk 3]
```diff
@@ -137,10 +137,8 @@ static RegistryAuthenticator fromAuthenticationMethod(
 
     Pattern servicePattern = Pattern.compile("service=\"(.*?)\"");
     Matcher serviceMatcher = servicePattern.matcher(authenticationMethod);
-    if (!serviceMatcher.find()) {
-      throw newRegistryAuthenticationFailedException(authenticationMethod, "service");
-    }
-    String service = serviceMatcher.group(1);
+    // use the provided registry as the service when missing (e.g., for OpenShift)
+    String service = serviceMatcher.find() ? serviceMatcher.group(1) : repository;
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java]: Oh, I meant that it should be like:

```
String service = serviceMatcher.find() ? serviceMatcher.group(1) : serverUrl;
```

Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java]: ie. `repository` is the image name

**Pull request merged**